distributed/dashboard/components/scheduler.py

```diff
@@ -1720,6 +1724,262 @@ def __del__(self):
+                }
+
+                # get dependents per task group
+                dependents = {k: [] for k in dependencies}
+                for k, v in dependencies.items():
+                    for dep in v:
+                        dependents[dep].append(k)
+
+                stack_order = toposort_layers(dependencies)
+                stack_it = stack_order[::-1].copy()
+
+                x = {}
+                y = {}
+                y_next = 0
+                collision = {}
+
+                # start and ends based on kwys and dependencies
+                arr_start = {}
+                arr_end = {}
+
+                nodes_layout = {}
+                arrows_layout = {}
+
+                while stack_it:
+                    tg = stack_it.pop()
+
+                    if not dependencies[tg]:
+                        x[tg] = 0
+                        y[tg] = y_next
+                        y_next += 1
+                    else:
+                        x[tg] = max(x[dep] for dep in dependencies[tg]) + 1 + 5
+
+                    # Given a task group I compute it's y position and it's dependants y-pos
+                    sort_dependents = [
+                        ele for ele in stack_order if ele in dependents[tg]
+                    ]
+
+                    for dep in sort_dependents:
+                        if dep in y:
+                            continue
+                        else:
+                            y[dep] = y[tg] + sort_dependents.index(dep)
+
+                    if (x[tg], y[tg]) in collision:
+
+                        old_x, old_y = x[tg], y[tg]
+                        x[tg], y[tg] = collision[(x[tg], y[tg])]
+
+                        y[tg] += 0.75  ##need to change when changing size of squares.
+                        collision[old_x, old_y] = (x[tg], y[tg])
+                    else:
+                        collision[(x[tg], y[tg])] = (x[tg], y[tg])
+
+                    # info neded for node layout to coulmn data source
+                    nodes_layout[tg] = {}
+                    nodes_layout[tg]["x"] = x[tg]
+                    nodes_layout[tg]["y"] = y[tg]
+
+                    # compute arrow layout info needed for column data source
+                    arr_start[tg] = dependencies[tg]
+                    arr_end[tg] = [tg] * len(dependencies[tg])
+
+                    # info needed for arrow layout
+                    arrows_layout[tg] = {}
+                    arrows_layout[tg]["nstart"] = arr_start[tg]
+                    arrows_layout[tg]["nend"] = arr_end[tg]
+
+                return nodes_layout, arrows_layout
+
+    @without_property_validation
+    def update(self):
+
+        if not self.scheduler.task_groups:
+            self.subtitle.text = "Scheduler is empty."
+        else:
+            self.subtitle.text = " "
+
+        if set(self.nodes_layout) != set(self.scheduler.task_groups):
```

Comment:
Dictionary key views support set-like operations, so we can drop the extra call to `set(...)` here

```suggestion
        if self.nodes_layout.keys() != self.scheduler.task_groups.keys():
```